src/cyrsasl_external.erl

```diff
@@ -0,0 +1,61 @@
+-module(cyrsasl_external).
+-author('denys.gonchar@erlang-solutions.com').
+
+-xep([{xep, 178}, {version, "1.1"}, {comment, "partially implemented."}]).
+
+-export([start/1, stop/0, mech_new/2, mech_step/2]).
+
+-behaviour(cyrsasl).
+
+-record(state, {creds}).
+
+start(_Opts) ->
+    cyrsasl:register_mechanism(<<"EXTERNAL">>, ?MODULE, cert),
+    ok.
+
+stop() ->
+    ok.
+
+-spec mech_new(Host :: ejabberd:server(),
+               Creds :: mongoose_credentials:t()) -> {ok, tuple()}.
+mech_new(_Host, Creds) ->
+  case mongoose_credentials:get(Creds,client_cert,undefined) of
```

Comment:
I would like to know on which step we add `client_cert` when reading the code.

Reply:
if client cert is missing, client can try another SASL mechanisms, it's not proposed to use SASL EXTERNAL.

Reply:
Is it possible for a client to reach `cyrsasl_external:new` without certificate present anyway?

Reply:
@arcusfelis 
It's pretty intuitive that the certificate is provided by c2s. And line 36 is not that far and it explains everything.

Reply:
no, it's not possible, see ejabberd_c2s, it filters out external mechanism if clients certificate is not provided.

Reply:
So this `case` is not necessary, correct? :)

Reply:
technically speaking - yes, but i don't see any valid reason to remove it. do you?

Reply:
Is it a good idea to introduce unreachable code?

Reply:
i don't see any harm, and this behavior corresponds to XEP

Reply:
Which part of the XEP covers a case when the EXTERNAL mechanism is offered but client hasn't provided any certificate?

Reply:
ok, you right, it's not [XEP178](https://xmpp.org/extensions/xep-0178.html), it's [RFC6120](https://tools.ietf.org/html/rfc6120#section-6.3.4). see ```6.3.4.  Mechanism Offers```

Reply:
`However, the receiving entity MAY offer the SASL EXTERNAL mechanism under other circumstances, as well.`

But we do not offer it under other circumstances, right? Currently it's an unreachable code that confuses the reader. YAGNI.

Reply:
it's quite fun to so YAGNI reference in a context of 1 extra code line, any way, SASL mechanizm filtering is not mandatory, and it simply was added later than was

Reply:
Is it important that filtering was added by this PR? The fact remains, that this code will never be executed unless there is a bug which would lead to client successfully choosing `EXTERNAL` despite lack of certificate.